api/client/push.go

```diff
@@ -42,6 +50,136 @@ func (cli *DockerCli) CmdPush(args ...string) error {
 	v := url.Values{}
 	v.Set("tag", tag)
 
-	_, _, err = cli.clientRequestAttemptLogin("POST", "/images/"+remote+"/push?"+v.Encode(), nil, cli.out, repoInfo.Index, "push")
-	return err
+	streamOut := cli.out
+	targets := []target{}
+	trustedPush := isTrusted(remote)
+	if trustedPush {
+		// Create multiWriter
+		r, w := io.Pipe()
+		defer w.Close()
+		// Capture output
+		streamOut = io.MultiWriter(streamOut, w)
+
+		go func() {
+			scanner := bufio.NewScanner(r)
+			scanner.Split(ScanInput)
+			// New splitter
+			for scanner.Scan() {
+				line := scanner.Text()
+				if matches := targetRegexp.FindStringSubmatch(line); len(matches) == 4 {
+					dgst, err := digest.ParseDigest(matches[2])
+					if err != nil {
+						fmt.Fprintf(cli.out, "Bad digest value %q, ignoring\n", matches[2])
+						continue
+					}
+					s, err := strconv.ParseInt(matches[3], 10, 64)
+					if err != nil {
+						fmt.Fprintf(cli.out, "Bad size value %q, ignoring\n", matches[3])
+						continue
+					}
+
+					targets = append(targets, target{
+						reference: registry.ParseReference(matches[1]),
+						digest:    dgst,
+						size:      s,
+					})
+				}
+			}
+
+		}()
+	}
+
+	_, _, err = cli.clientRequestAttemptLogin("POST", "/images/"+remote+"/push?"+v.Encode(), nil, streamOut, repoInfo.Index, "push")
+	if err != nil {
+		return err
+	}
+	if !trustedPush {
+		return nil
+	}
+
+	if tag == "" {
```

Comment:
Push has the confusing behavior of treating `tag == ""` as push all tags while pull treats it as pull latest.